src/core_plugins/kibana/public/context/app.js

```diff
@@ -0,0 +1,103 @@
+import _ from 'lodash';
+
+import uiModules from 'ui/modules';
+import contextAppTemplate from './app.html';
+import './components/loading_button';
+import './components/size_picker/size_picker';
+import { bindSelectors } from './utils/selectors';
+import {
+  QueryParameterActionsProvider,
+  QUERY_PARAMETER_KEYS,
+} from './query_parameters';
+import {
+  QueryActionsProvider,
+  selectIsLoadingAnchorRow,
+  selectIsLoadingPredecessorRows,
+  selectIsLoadingSuccessorRows,
+  selectRows,
+} from './query';
+
+const module = uiModules.get('apps/context', [
+  'kibana',
+  'ngRoute',
+]);
+
+module.directive('contextApp', function ContextApp() {
+  return {
+    bindToController: true,
+    controller: ContextAppController,
+    controllerAs: 'contextApp',
+    restrict: 'E',
+    scope: {
+      anchorUid: '=',
+      columns: '=',
+      indexPattern: '=',
+      predecessorCount: '=',
+      successorCount: '=',
+      sort: '=',
+    },
+    template: contextAppTemplate,
+  };
+});
+
+function ContextAppController($scope, Private) {
+  const queryParameterActions = Private(QueryParameterActionsProvider);
+  const queryActions = Private(QueryActionsProvider);
+
+  this.state = createInitialState();
+
+  this.actions = _.mapValues({
+    ...queryParameterActions,
+    ...queryActions,
+  }, (action) => (...args) => action(this.state)(...args));
+
+  this.derivedState = bindSelectors({
+    isLoadingAnchorRow: selectIsLoadingAnchorRow,
+    isLoadingPredecessorRows: selectIsLoadingPredecessorRows,
+    isLoadingSuccessorRows: selectIsLoadingSuccessorRows,
+    rows: selectRows,
+  }, () => this.state);
+
+  /**
+   * Sync query parameters to arguments
+   */
+  $scope.$watchCollection(
+    () => _.pick(this, QUERY_PARAMETER_KEYS),
+    (newValues) => {
+      // break the watch cycle
```

Comment:
The problem with the flow you are proposing lies in step 2: The change was originally caused by something else calling `fetchAllRowsWithNewQueryParameters`. So this action will be dispatched a second time causing another request being sent to Elasticsearch. This `$watchCollection` is just meant to be triggered if the query parameters are changed from outside this directive. Any change triggered from the inside has already been handled and this is just feedback from the synchronization performed by the other `$watchCollection` below.